shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/TypeFamilyClasses.hs

```diff
@@ -0,0 +1,335 @@
+
+{------------------------------------------------------------------------------
+  We gather together all the operations on the type family Script that are
+  necessary to generate consistent transactions to run property tests.
+  The key component is the class ScriptClass.
+ -----------------------------------------------------------------------------}
+module Test.Shelley.Spec.Ledger.Generator.TypeFamilyClasses
+  ( ScriptClass (..),
+    Quantifier (..),
+    ValueClass (..),
+    TxBodyClass (..),
+    genCoin,
+    exponential,
+    anyOf,
+    allOf,
+    mOf,
+    ScriptPairs,
+    keyPairs,
+    mkPayScriptHashMap,
+    mkStakeScriptHashMap,
+    mkScriptsFromKeyPair,
+    mkKeyPairs,
+    mkScripts,
+    mkScriptCombinations,
+    combinedScripts,
+    someScripts,
+    scriptKeyCombinations,
+    scriptKeyCombination,
+    genesisId,
+  )
+where
+
+import Cardano.Binary (Annotator, FromCBOR (..), ToCBOR (..))
+-- also to import instances
+
+import Cardano.Crypto.DSIGN.Class (DSIGNAlgorithm (..))
+import qualified Cardano.Crypto.Hash as Hash
+import Cardano.Ledger.Compactible (CompactForm (..), Compactible (..))
+import Cardano.Ledger.Core (AnnotatedData, ChainData, Script, SerialisableData, TxBody, Value)
+import qualified Cardano.Ledger.Core as Core
+import Cardano.Ledger.Crypto (DSIGN)
+import qualified Cardano.Ledger.Crypto as CC (Crypto)
+import Cardano.Ledger.Era (Era (..))
+import Cardano.Ledger.Shelley (ShelleyEra)
+import Cardano.Ledger.Torsor (Torsor (..))
+import Cardano.Ledger.Val (Val (..))
+import Data.List (concatMap, permutations)
+import qualified Data.List as List
+import qualified Data.Map as Map
+import Data.Proxy
+import qualified Data.Sequence.Strict as StrictSeq
+import qualified Data.Set as Set
+import Data.Tuple (swap)
+import Data.Word (Word64)
+import NoThunks.Class (NoThunks (..))
+import Shelley.Spec.Ledger.BaseTypes (StrictMaybe (..))
+import Shelley.Spec.Ledger.Coin (Coin (..))
+import Shelley.Spec.Ledger.Hashing (EraIndependentTxBody, HashAnnotated (..))
+import Shelley.Spec.Ledger.Keys (Hash, KeyHash, KeyPair (..), KeyRole (..), VKey (..), asWitness, hashKey, vKey)
+import Shelley.Spec.Ledger.LedgerState (KeyPairs)
+import Shelley.Spec.Ledger.Scripts (MultiSig (..), ScriptHash, hashMultiSigScript)
+import Shelley.Spec.Ledger.Slot (SlotNo (..))
+import Shelley.Spec.Ledger.Tx (ValidateScript (..))
+import Shelley.Spec.Ledger.TxBody (TxId (..))
+import qualified Shelley.Spec.Ledger.TxBody as Shelley
+import Test.QuickCheck (Gen, generate)
+import qualified Test.QuickCheck as QC
+import Test.Shelley.Spec.Ledger.Generator.Constants
+  ( Constants (..),
+    defaultConstants,
+  )
+import Test.Shelley.Spec.Ledger.Utils (mkKeyPair)
+
+class
+  ( Show (Script era),
+    Eq (Script era),
+    ValidateScript era,
+    HashIndex (Core.TxBody era) ~ EraIndependentTxBody,
+    Eq (Core.TxBody era),
+    Show (Core.TxBody era),
+    CC.Crypto (Crypto era),
+    Era era
+  ) =>
+  ScriptClass era
```

Comment:
I think `ScriptClass` (and associated functions) is big enough to deserve its own module

Reply:
Generator/Scripts?